src/sources/alphavantage.rs

```diff
@@ -0,0 +1,152 @@
+//! Alphavantage Source Provider
+//! <https://www.alphavantage.co/>
+//!
+
+use super::{Pair, Price};
+use crate::error::Error;
+use bytes::buf::ext::BufExt;
+use hyper::{
+    client::{Client, HttpConnector},
+    header, Body, Request,
+};
+use hyper_rustls::HttpsConnector;
+use serde::{Deserialize, Serialize};
+
+/// Base URI for requests to the Coinone API
+pub const BASE_URI: &str = "https://www.alphavantage.co/";
+
+/// User-Agent to send in HTTP request
+pub const USER_AGENT: &str = "iqlusion delphi";
+
+/// Source provider for Alphavantage
+pub struct AlphavantageSource {
+    http_client: Client<HttpsConnector<HttpConnector>>,
+}
+
+///Parameters for queries
+pub struct AlphavantageParams {
+    function: String,
+    from_currency: String,
+    to_currency: String,
+    apikey: String,
+}
+
+impl AlphavantageParams {
+    ///Convert params into url query parameters
+    pub fn to_query_string(&self) -> String {
+        format!(
+            "function={}&from_currency={}&to_currency={}&apikey={}",
+            self.function, self.from_currency, self.to_currency, self.apikey
+        )
+    }
```

Comment:
To avoid string concatenation issues/escaping bugs if this were to change, you could write this in a bit more extensible/type-safe way by doing something similar to how [`reqwest::RequestBuilder::query`](https://docs.rs/reqwest/0.10.8/reqwest/struct.RequestBuilder.html#method.query) works, and construct a `Uri` type in the process:

```suggestion
    pub fn to_request_uri(&self) -> Uri {
        let query = [
            ("function", &self.function),
            ("from_currency", &self.from_currency),
            ("to_currency", &self.to_currency),
            ("apikey", &self.apikey)
        ].iter()
         .map(|(k,v)| format!("{}={}", k, v))
         .collect::<Vec<_>>()
         .join("&");
    
    let path_and_query = format!("/{}", query);
    
    uri::Builder::new()
        .scheme("https")
        .authority("www.alphavantage.co")
        .path_and_query(path_and_query.as_str())
        .build()
    }
```

This would also make it easier to move to `reqwest` in the future if we decide to do that.

Reply:
@poldsam FYI URL encoding https://www.w3schools.com/tags/ref_urlencode.ASP